### PR TITLE
Weapon Category Update

### PR DESCRIPTION
**Pull request by KheirFerrum**
#### Summary

SUMMARY: Infrastructure "Port over Weapon Categories and tweak to BN specifications."

#### Purpose of change

Add Weapon Categories, which allow easier addition of weapons to martial arts and hopefully in the future can be used to aid with holsters and magazines, to define what can and cannot be stored without writing a full thesis.

#### Describe the solution

Ports over all related PRs to this [issue](https://github.com/CleverRaven/Cataclysm-DDA/issues/51867).

We will need to tweak these. Because from what I can read the system was updated at some point to force weapon categories, which honestly isn't the kind of system I'm looking for.

Category Display has been added with a rather different implementation from the [inspiring](https://github.com/CleverRaven/Cataclysm-DDA/pull/52638) PR, now the game will display the weapon categories listed in the MA jsons along with all applicable weapons. If you add a weapon via the old `"weapons"` variable it will list it as "Special" weapons for that style.

Category Translation was merged. So now it's not screaming in all caps and _.

Contains a fix for techniques. Previously the game would only read and add techniques, never clearing them when using copy-from in any form. Now it properly inherits and clears. I have not personally tested extend/delete.

#### Describe alternatives you've considered

Lots of FLAGs.

#### Testing

- [x] Make sure all weapons including guns are properly categorized.
- [x] Make sure that weapon styles display correctly.
- [x] Add in [Category Display](https://github.com/CleverRaven/Cataclysm-DDA/pull/52638) or equivalent.
- [x] Make sure I didn't accidentally add the fragments of any _other_ implementations people were doing throughout this issue.

#### Additional context

PRs like [Category Display](https://github.com/CleverRaven/Cataclysm-DDA/pull/52638), [Category Translation](https://github.com/CleverRaven/Cataclysm-DDA/pull/52639) could probably be merged, with fixes, mostly because the category display is part of the reason weapon categories have been enforced.

image

### CODE REVIEW

**Comment by Coolthulhu** [comment]
Is this ready or still a draft?

**Comment by KheirFerrum** [comment]
Gave up on making it a single category only. Super-categories are not on my agenda for this PR, though I think it should easily be possible.

File: doc/JSON_INFO.md [hunk 1]
```diff
@@ -2666,6 +2744,18 @@ For every `type` other then `bionic` and `bionic_group` following entries scale
...
+```c++
+{
+    "type": "weapon_category",
+    "id": "WEAP_CAT"
+    "name": "Weapon Category"
+}
+```
+
+`"name"` is a translatable string used for UI display in martial arts UI, while ID is used for JSON entries.
+
```
Comment (olanti-p) [inline doc/JSON_INFO.md:2758]: Maybe add a short note describing what weapon categories are for?

File: lang/extract_json_strings.py [hunk 2]
```diff
@@ -807,6 +807,12 @@ def extract_snippet(state, item):
         else:
             writestr(state, snip["text"])
 
+def extract_weapon_category(item):
+    outfile = get_outfile("weapon_category")
+    name = item.get("name")
+    comment = "weapon category name"
+    writestr(outfile, name, comment=comment)
+
```
Comment (olanti-p) [inline lang/extract_json_strings.py:815]: This function must be registered in `extract_specials` table on line 818

File: src/item.cpp [hunk 3]
```diff
@@ -1508,6 +1508,14 @@ void item::basic_info( std::vector<iteminfo> &info, const iteminfo_query *parts,
         info.push_back( iteminfo( "BASE", _( "Category: " ),
                                   "<header>" + get_category().name() + "</header>" ) );
     }
+    if( !type->weapon_category.empty() && parts->test( iteminfo_parts::WEAPON_CATEGORY ) ) {
+        const std::string weapon_categories = enumerate_as_string( type->weapon_category.begin(),
+        type->weapon_category.end(), [&]( const weapon_category_id & elem ) {
+            return elem->name().translated();
+        }, enumeration_conjunction::none );
+        info.push_back( iteminfo( "BASE", _( "Weapon Categories: " ),
```
Comment (olanti-p) [inline src/item.cpp:1516]: I think singular would be better here for consistency, UI uses singular for nearby `Category` and `Material` entries (even when there are multiple materials)
```suggestion
        info.push_back( iteminfo( "BASE", _( "Weapon Category: " ),
```

File: src/martialarts.cpp [hunk 4]
```diff
@@ -1483,26 +1529,68 @@ bool ma_style_callback::key( const input_context &ctxt, const input_event &event
...
+                                             c_light_cyan ) :
+                                             carried ? colorize( item::nname( wid ), c_yellow ) : item::nname( wid );
+                    weapons.push_back( weaponname );
+                }
+                // Sorting alphabetically makes it easier to find a specific weapon
+                std::sort( weapons.begin(), weapons.end(), localized_compare );
+                // This removes duplicate names (e.g. a real weapon and a replica sharing the same name) from the weapon list.
+                auto last = std::unique( weapons.begin(), weapons.end() );
+                weapons.erase( last, weapons.end() );
+                buffer += std::string( "<header>Special: </header>" );
```
Comment (olanti-p) [inline src/martialarts.cpp:1591]: Missing translation
```suggestion
                buffer += _( "<header>Special: </header>" );
```

**Review by olanti-p — COMMENTED** [review]
I didn't go over all the items and techniques, but the rest is good, only a couple minor issues.

File: doc/JSON_INFO.md (hunk 1 above)
Comment (KheirFerrum) [reply doc/JSON_INFO.md]: Not sure if there's anything else to add for now, but I gave it a brief description

File: src/item_factory.cpp [hunk 5]
```diff
@@ -2438,6 +2438,16 @@ void Item_factory::load_basic_info( const JsonObject &jo, itype &def, const std:
...
     }
 
+    if( jo.has_member( "weapon_category" ) ) {
+        optional( jo, true, "weapon_category", def.weapon_category, auto_flags_reader<weapon_category_id> {} );
+        for( const weapon_category_id &cat_id : def.weapon_category ) {
+            if( !cat_id.is_valid() ) {
+                jo.throw_error( string_format( "invalid weapon category: %s", std::string( cat_id ) ),
+                                "weapon categories" );
+            }
+        }
```
Comment (olanti-p) [inline src/item_factory.cpp:2448]: This check should be moved to `Item_factory::check_definitions()`.

Data loading is done in multiple stages, and at this stage not all definitions has been loaded from JSON yet. There's no guarantee whether the game has loaded the file where the weapon categories are defined, and so may throw a false positive.

So, such things are usually checked in either consistency or finalization stages - after the game has loaded all JSONs and resolved copy-froms, so it's possible to check the ids, but before the data is used for a game world.

File: lang/extract_json_strings.py [hunk 6]
```diff
@@ -807,6 +807,12 @@ def extract_snippet(state, item):
         else:
             writestr(state, snip["text"])
 
+def extract_weapon_category(item):
+    outfile = get_outfile("weapon_category")
+    name = item.get("name")
+    comment = "weapon category name"
+    writestr(outfile, name, comment=comment)
```
Comment (olanti-p) [inline lang/extract_json_strings.py:814]: The script has been modified in BN, and the extraction functions use a different output algorithm.

```suggestion

def extract_weapon_category(state, item):
    name = item["name"]
    comment = "weapon category name"
    writestr(state, name, comment=comment)
```

File: lang/extract_json_strings.py [hunk 7]
```diff
@@ -822,6 +828,7 @@ def extract_snippet(state, item):
     "harvest" : extract_harvest,
     "mapgen": extract_mapgen,
     "martial_art": extract_martial_art,
+    "weapon_category": extract_weapon_category,
```
Comment (olanti-p) [inline lang/extract_json_strings.py:831]: Please keep the list alphabetic.

File: src/init.cpp [hunk 8]
```diff
@@ -547,6 +548,7 @@ void DynamicDataLoader::unload_data()
     bionic_data::reset();
     body_part_type::reset();
     charge_removal_blacklist::reset();
+    weapon_category::reset();
```
Comment (olanti-p) [inline src/init.cpp:551]: Alphabetic, plz

**Review by olanti-p — COMMENTED** [review]
I didn't check the melee weapons yet, but the ranged ones are ok.

File: data/json/items/classes/gun.json [hunk 9]
```diff
@@ -216,10 +279,92 @@
...
+    "weapon_category": [ "RIFLES", "MAGNETIC" ],
+    "name": { "str": "magnetic rifle" },
+    "looks_like": "laser_rifle",
+    "//": "No conventional faults because it uses non-conventional systems.",
+    "flags": [ "NEVER_JAMS", "NON-FOULING", "NEEDS_NO_LUBE" ],
+    "valid_mod_locations": [
+      [ "accessories", 4 ],
+      [ "emitter", 1 ],
+      [ "grip", 1 ],
+      [ "lens", 1 ],
```
[the remark this answers is not in the document]
Comment (KheirFerrum) [reply data/json/items/classes/gun.json:317]: I fucked up and didn't copy the actual magnetic weapon mod locations over.

**Comment by olanti-p** [comment]
Skimmed the melee weapons, look fine.
The last thing missing is weapon category validness check in `check_martialarts()` and I think it will be good to go.

**Review by olanti-p — APPROVED** [review]

**Pull request merged**